### PR TITLE
[python-package] feature_names_in_ spaces replaced by underlines incompatible with sklearn

### PR DESCRIPTION
**Pull request by ThibaultDECO**
Fixes issue #7338

FIX feature_names_in_ whitespace replaced by underlines is incompatible with sklearn

### CODE REVIEW

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for your interest in LightGBM and for taking the time to open a PR.

I've started CI for you without reviewing the changes closely yet, let's see what that says.

At a glance, I'll say I'm not very excited about having to track the original feature names everywhere. This is new complexity and that means new opportunity for bugs and maintenance burden 😫 

I will review more closely once we've release v4.7.0 (#7129), that is my primary focus right now.

**Comment by jameslamb** [comment]
Because you're a first-time contributor, every CI run requires one of us to click a button. You may find it faster to run the tests locally.

```shell
cmake -B build -S .
cmake --build build --target _lightgbm -4
sh build-python.sh install --precompile
pytest tests/python_package_test/
```

**Comment by ThibaultDECO** [comment]
@jameslamb Thank you James, I will run the tests locally.
With regards to your pushback, I understand that it adds complexity but I don't see any other way around at this time.
Thank you for your review!

**Comment by jameslamb** [comment]
> _I don't see any other way around at this time._

Other options might be:

* raise an informative error closer to the boundary of the `scikit-learn` interface if any feature names have spaces (i.e. simply don't allow them)
* remove the normalization of feature names completely in `Dataset` construction (i.e. allow spaces)
* make the feature name normalization configurable via a new parameter, and have `scikit-learn` interface explicitly opt out of that when it constructs a `Dataset`
* stop supporting `feature_names_in_` in the scikit-learn interface entirely (maybe it is just not worth it, we keep tripping over it here)

I think evaluating this will require coming up with a specific answer for the question "why does LightGBM remove spaces from feature names at all?", which will require looking through some project history.
